File: algorithms/two_pointers/11-container_with_most_water.go
package main

// 2023-04-27
// TC: O(n) - Iterating through height arr using two pointers
// SC: O(n) - Pointer logic no extra space needed
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func maxArea(height []int) int {
	max := 0
	l := 0
	r := len(height) - 1
	for l < r {
		// Calculate area
		a := min(height[l], height[r]) * (r - l)
		if a > max {
			max = a
		}

		if height[l] < height[r] {
			l++
		} else {
			r--
		}
	}

	return max
}

func main() {
}

File: algorithms/python-algorithms/two_pointers/11-container_with_most_water.py
from typing import List


# 2023-05-16
# TC: O(n) - Use two pointers to iterate through entire arr
# SC: O(1) - Pointer logic only no extra space
class Solution:
    def maxArea(self, height: List[int]) -> int:
        res = 0
        lo, hi = 0, len(height) - 1

        while lo < hi:
            area = (hi - lo) * min(height[lo], height[hi])
            res = max(res, area)

            if height[lo] < height[hi]:
                lo += 1
            else:
                hi -= 1

        return res

    def maxAreaBruteForce(self, height: List[int]) -> int:
        res = 0

        for lo in range(len(height)):
            for hi in range(lo + 1, len(height)):
                area = (hi - lo) * min(height[lo], height[hi])
                res = max(res, area)

        return res


def main():
    pass


if __name__ == "__main__":
    main()
